In [50]:
#importing libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.compose import ColumnTransformer
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline,make_pipeline
from sklearn.tree import DecisionTreeClassifier


In [51]:
#loading the dataset
df = pd.read_csv('train.xls')

#day 19 learnings
1) how big data ?

In [52]:
df.shape#so our data has 12 columns and around 891 rows 

(891, 12)

2).how data looks like?

In [53]:
df.sample(8)

PassengerId  Survived  Pclass                             Name     Sex  \
611          612         0       3            Jardin, Mr. Jose Neto    male   
368          369         1       3              Jermyn, Miss. Annie  female   
762          763         1       3            Barah, Mr. Hanna Assi    male   
302          303         0       3  Johnson, Mr. William Cahoone Jr    male   
729          730         0       3    Ilmakangas, Miss. Pieta Sofia  female   
832          833         0       3                   Saad, Mr. Amin    male   
483          484         1       3           Turkula, Mrs. (Hedwig)  female   
174          175         0       1          Smith, Mr. James Clinch    male   

      Age  SibSp  Parch              Ticket     Fare Cabin Embarked  
611   NaN      0      0  SOTON/O.Q. 3101305   7.0500   NaN        S  
368   NaN      0      0               14313   7.7500   NaN        Q  
762  20.0      0      0                2663   7.2292   NaN        C  
302  19.0      0      0                LINE   0.0000   NaN        S  
729  25.0      1      0    STON/O2. 3101271   7.9250   NaN        S  
832   NaN      0      0                2671   7.2292   NaN        C  
483  63.0      0      0                4134   9.5875   NaN        S  
174  56.0      0      0               17764  30.6958    A7        C

In [54]:
df.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [55]:
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

inference=>1)Age has missing values 177 
           2)Cabin has 687 missing values out of 891 rows so we will drop this
           3)Embarked has 2 missing values

In [56]:
df.duplicated().sum()#no duplicate values

0

In [57]:
# df['Family']=df['SibSp']+df['Parch']
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [58]:
df.drop(columns=['PassengerId','Name','Ticket','Cabin'],inplace=True)#removing columns Passengerid,Name,Cabin

In [59]:
X_train,X_test,y_train,y_test = train_test_split(df.drop(columns=['Survived']),
                                                 df['Survived'],
                                                  test_size=0.2,
                                                random_state=42)
X_train

# df = pd.DataFrame(trns2.fit_transform(X_train))
# print(df)

Pclass     Sex   Age  SibSp  Parch      Fare Embarked
331       1    male  45.5      0      0   28.5000        S
733       2    male  23.0      0      0   13.0000        S
382       3    male  32.0      0      0    7.9250        S
704       3    male  26.0      1      0    7.8542        S
813       3  female   6.0      4      2   31.2750        S
..      ...     ...   ...    ...    ...       ...      ...
106       3  female  21.0      0      0    7.6500        S
270       1    male   NaN      0      0   31.0000        S
860       3    male  41.0      2      0   14.1083        S
435       1  female  14.0      1      2  120.0000        S
102       1    male  21.0      0      1   77.2875        S

[712 rows x 7 columns]

#we will do eda later in this part but first make a pipeline so to revise the steps:
1)impute the missing values 
2)feature construction
3)handle categorical features
4)feature construction
5)feature scaling
6)model training



In [60]:
# trns1=ColumnTransformer([('age_impute',SimpleImputer(strategy='mean'),[3]),
#                         ('embark_impute',SimpleImputer(strategy='most_frequent'),[5])],
#                         remainder='passthrough')

trns1 = ColumnTransformer([
    ('age_impute', SimpleImputer(), [2]),
    ('embark_impute', SimpleImputer(strategy='most_frequent'), [6])
], remainder='passthrough')




In [61]:
# trns2=ColumnTransformer([
#     ('ohe_sex_embarked',OneHotEncoder(sparse_output=False,drop='first',handle_unknown='ignore'),[1,6])
# ],remainder='passthrough')
trns2 = ColumnTransformer([
    ('ohe_sex_embarked',OneHotEncoder(sparse=False,handle_unknown='ignore'),[1,6])
],remainder='passthrough')
# trns2.fit(df)

# Get the number of output columns
# num_columns = trns2.transform(df).shape[1]
# num_columns = trns2.transform(df)
# num_columns
# print("Number of columns after transformation:", num_columns)#want to know no. of columns to apply scaling

In [62]:
trns3=ColumnTransformer([('Scaling',StandardScaler(),slice(0,None))])# Scaling all columns

In [63]:
trns4=DecisionTreeClassifier()

Creating pipeline

In [64]:
pipe=Pipeline([
   ( 'trns1',trns1),
     ('trns2',trns2)
#     ('trns3',trns3),
#     ('trns4',trns4)
])

In [65]:
pipe.fit(X_train,y_train)

# If you're doing supervised learning, you would use 'y_train' during model training
# pipe.fit(X_train, y_train)

# Print the transformed data


D:\ana\Lib\site-packages\sklearn\preprocessing\_encoders.py:972: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


Pipeline(steps=[('trns1',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('age_impute', SimpleImputer(),
                                                  [2]),
                                                 ('embark_impute',
                                                  SimpleImputer(strategy='most_frequent'),
                                                  [6])])),
                ('trns2',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('ohe_sex_embarked',
                                                  OneHotEncoder(handle_unknown='ignore',
                                                                sparse=False),
                                                  [1, 6])]))])

In [66]:
X_train


Pclass     Sex   Age  SibSp  Parch      Fare Embarked
331       1    male  45.5      0      0   28.5000        S
733       2    male  23.0      0      0   13.0000        S
382       3    male  32.0      0      0    7.9250        S
704       3    male  26.0      1      0    7.8542        S
813       3  female   6.0      4      2   31.2750        S
..      ...     ...   ...    ...    ...       ...      ...
106       3  female  21.0      0      0    7.6500        S
270       1    male   NaN      0      0   31.0000        S
860       3    male  41.0      2      0   14.1083        S
435       1  female  14.0      1      2  120.0000        S
102       1    male  21.0      0      1   77.2875        S

[712 rows x 7 columns]